Library/Homebrew/formula_printer.rb

```diff
@@ -0,0 +1,72 @@
+require "formula"
+require "tab"
+require "diagnostic"
+
+module Homebrew
+  class FormulaPrinter
+    attr_reader :domain, :selection
+
+    def initialize
+      @domain = domain
+      @selection = selection
+    end
+
+    def selection
+      raise "method missing"
+    end
+
+    def domain
+      if ARGV.named.empty?
+        Formula.installed
+      else
+        ARGV.resolved_formulae
+      end
+    end
+
+    def self.print
+      return unless HOMEBREW_CELLAR.exist?
+      printer = new
+      if ARGV.json == "v1"
+        printer.print_json
+      elsif ($stdout.tty? || ARGV.verbose?) && !ARGV.flag?("--quiet")
+        printer.print_verbose
+      else
+        printer.print_quiet
+      end
+      printer.exit_status
+    end
+
+    def exit_status
+      Homebrew.failed = @selection.any?
+    end
+
+    def quiet
+      @selection.class == Hash ? @selection.values.flatten.uniq : @selection
```

Comment:
Alternatively, it could be either
- abstract and always implemented in subclass

or

- defined here as `@subset` by default, and then overridden in `missing` as `@subset.values.flatten.uniq`